File: backend/src/main/java/com/ironpath/backend/bodymetrics/domain/model/BodyMeasurement.java
package com.ironpath.backend.bodymetrics.domain.model;

import com.ironpath.backend.identity.domain.model.User;
import com.ironpath.backend.shared.infrastructure.EncryptedDoubleConverter;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "body_measurements")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class BodyMeasurement {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(columnDefinition = "TEXT")
    private Double weight;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(columnDefinition = "TEXT")
    private Double chest;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(columnDefinition = "TEXT")
    private Double waist;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(columnDefinition = "TEXT")
    private Double hips;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(name = "left_arm", columnDefinition = "TEXT")
    private Double leftArm;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(name = "right_arm", columnDefinition = "TEXT")
    private Double rightArm;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(name = "left_thigh", columnDefinition = "TEXT")
    private Double leftThigh;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(name = "right_thigh", columnDefinition = "TEXT")
    private Double rightThigh;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(name = "left_calf", columnDefinition = "TEXT")
    private Double leftCalf;

    @Convert(converter = EncryptedDoubleConverter.class)
    @Column(name = "right_calf", columnDefinition = "TEXT")
    private Double rightCalf;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "recorded_at", nullable = false)
    private LocalDateTime recordedAt;

    @Column(name = "archived")
    private Boolean archived = false;

    @PrePersist
    protected void onCreate() {
        if (recordedAt == null) {
            recordedAt = LocalDateTime.now();
        }
    }
}

File: frontend/lib/features/bodymetrics/domain/models/body_measurement.dart
class BodyMeasurement {
  final String id;
  final double? weight;
  final double? chest;
  final double? waist;
  final double? hips;
  final double? leftArm;
  final double? rightArm;
  final double? leftThigh;
  final double? rightThigh;
  final double? leftCalf;
  final double? rightCalf;
  final String? notes;
  final DateTime recordedAt;
  final bool archived;

  BodyMeasurement({
    required this.id,
    this.weight,
    this.chest,
    this.waist,
    this.hips,
    this.leftArm,
    this.rightArm,
    this.leftThigh,
    this.rightThigh,
    this.leftCalf,
    this.rightCalf,
    this.notes,
    required this.recordedAt,
    this.archived = false,
  });

  bool get isEditable => !archived;

  factory BodyMeasurement.fromJson(Map<String, dynamic> json) {
    return BodyMeasurement(
      id: json['id'],
      weight: json['weight']?.toDouble(),
      chest: json['chest']?.toDouble(),
      waist: json['waist']?.toDouble(),
      hips: json['hips']?.toDouble(),
      leftArm: json['leftArm']?.toDouble(),
      rightArm: json['rightArm']?.toDouble(),
      leftThigh: json['leftThigh']?.toDouble(),
      rightThigh: json['rightThigh']?.toDouble(),
      leftCalf: json['leftCalf']?.toDouble(),
      rightCalf: json['rightCalf']?.toDouble(),
      notes: json['notes'],
      recordedAt: DateTime.parse(json['recordedAt']),
      archived: json['archived'] ?? false,
    );
  }
}
